### ISSUE TITLE
Parameter substitution ignored by transpiler 

### ISSUE THREAD

**Issue by Cryoris** [Qiskit/qiskit-terra#3509]
### Information

- **Qiskit Terra version**: 0.11.0, commit bd19bb1e16b2dcf697af01f9b03ff61f826c05e6
- **Python version**: 3.7.5
- **Operating system**: MacOS Mojave (10.14.6)

### What is the current behavior?
In an `Instruction`, parameters of type `Parameter` may be substituted with another `Parameter` object using the `params` setter. This, however, changes only the member `_params` of the instruction, not the parameters of the gates inside the instruction. 
If the instruction is appended to a circuit, this can cause the following two issues:

1. if the circuit is decomposed and drawn, the parameter name has not been changed.
2. if the circuit is transpiled and basis gates are specified, the old parameter object is still in place. This does not happen if no basis gates are provided.

Both problems arise because on updating the parameters of an instruction, this update is not propagated to the gates (or sub-instructions) it contains.

### Steps to reproduce the problem

For problem case 1: 
```
qc = QuantumCircuit(1)
qc.ry(Parameter('original'), 0)
instr = qc.to_instruction()
instr.params = [Parameter('new')]

from_instr = QuantumCircuit(1)
from_instr.append(instr, [0])
print(from_instr.draw())
print(from_instr.decompose().draw())
```
yields the output
```
        ┌────────────────┐
q_0: |0>┤ Circuit52(new) ├
        └────────────────┘
        ┌──────────────┐
q_0: |0>┤ Ry(original) ├
        └──────────────┘
```

For problem case 2:
```
qc = QuantumCircuit(1)
qc.ry(Parameter('original'), 0)
instr = qc.to_instruction()
instr.params = [Parameter('new')]

from_instr = QuantumCircuit(1)
from_instr.append(instr, [0])

backend = Aer.get_backend("statevector_simulator")
basis_gates = backend.configuration().basis_gates
print("w/o backend or gates:", transpile(from_instr).parameters)
print("w/ backend:", transpile(from_instr, backend).parameters)
print("w/ basis gates:", transpile(from_instr, basis_gates=basis_gates).parameters)
```
results in
```
w/o backend or gates: {Parameter(new)}
w/ backend: {Parameter(original)}
w/ basis gates: {Parameter(original)}
```

### What is the expected behavior?

The expected behaviour is that setting the parameters of an instruction via `instr.params = [Parameter('new')]` overwrites the old parameters.

More concretely, the expected output for problem 2 is:
```
w/o backend or gates: {Parameter(new)}
w/ backend: {Parameter(new)}  # instead of Parameter(original)
w/ basis gates: {Parameter(new)}  # instead of Parmeter(original)
```

### Suggested solutions
On calling the setter of `Instruction.params`, propagate the updates to the gates/instructions in `Instruction.definition`.

**Comment by Cryoris** [comment]
This is probably the minimal example showing the problem
```
qc = QuantumCircuit(1)
qc.ry(Parameter('original'), 0)
instr = qc.to_instruction()

instr.params = [Parameter('new')]

print("instruction params:", instr.params)
print("ry params:", instr.definition[0][0].params)
```
results in
```
instruction params: [Parameter(new)]
ry params: [Parameter(original)]
```

**Comment by kdk** [comment]
Fixed by #3729 .
